BUGFIX: Ensure `Node::$timestamps` uses `UTC` format consistently instead server timezone

mhsdesign opened the pull request:
Requires https://github.com/neos/eventstore-doctrineadapter/pull/29

In the database we cant persist the timezone but only the date. ~And DBAL seems to automatically store them in UTC which makes sense.~ In DBAL we use `Doctrine\DBAL\Types\DateTimeImmutableType::convertToDatabaseValue()` which invokes only `->format` for the date as `Y-m-d H:i:s`. 

Now the date comes from https://github.com/neos/neos-development-collection/blob/c1a329f3a3f9e6c36e83484cf3d90d345a8e2169/Neos.ContentRepositoryRegistry/Classes/Factory/Clock/SystemClock.php#L15 (or the FakeClock)

And is written into the `recordedat` column of the event: https://github.com/neos/eventstore-doctrineadapter/blob/e7d35402e05280be82794fbe507062d0e6cf48e0/src/DoctrineEventStore.php#L304

Either we must ensure we use only UTC dates or we change the date to UTC in the event store.

Then we can merge this change which ensures that during deserializing we always use `UTC` and never ask the system `date_default_timezone_get ()` as this should be irrelevant.


<!-- 
    Thanks for your contribution, we appreciate it!

    The first section should explain briefly what is changed. 
    Some examples are always nice to showcase the use. 
    The content will be used in the change-logs and addresses 
    developers working with Neos.

    If there are issues regarding the topic of your PR link 
    them here as `related:` or `resolved:`
-->

**Upgrade instructions**

With the change https://github.com/neos/neos-development-collection/pull/5858
a new kind of migration tooling will be provided to upgrade Neos projects: Content repository upgrades.

These upgrades are available via `crupgrade` and contain for example like in the Neos 9 beta phase tooling to adjust and correct your events.

In the case of this UTC change if your server was previously configured in local PHP time and you consider the `recordedAt` dates really important for legal reasons in example, you can choose to fix these.

This is a fully optimal migration and which can also be executed once your upgrading to a next minor version like 9.2 

```
./flow crupgrade:eventsrecordedattoutc
```

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

**Review instructions**

<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mhsdesign committed:
BUGFIX: Ensure Node->timestamps use `UTC` format consistently and not `date_default_timezone_get()` base on the system

In the database we dont persist the timezone but only the date. And dbal seems to automatically store them in UTC which makes sense.

During deserializing we need to ensure that we use always `UTC` too and never ask the system for its fallback as this is irrelevant.

Commit:
WIP: UTC all the dates - store InitiatingTimestamp as UTC

mhsdesign committed:
TASK: Ensure originalCreated (with timezone) is stored as UTC

see

https://github.com/neos/neos-development-collection/pull/5716#issuecomment-4438715550

mhsdesign commented on Neos.ContentRepository.Core/Classes/EventStore/InitiatingEventMetadata.php:
```diff
@@ -34,7 +34,8 @@ public static function enrichEventsWithInitiatingMetadata(
         \DateTimeImmutable $initiatingTimestamp,
     ): Events {
-        $initiatingTimestampFormatted = $initiatingTimestamp->format(\DateTimeInterface::ATOM);
+        $initiatingTimestampFormatted = $initiatingTimestamp->setTimezone(new \DateTimeZone('UTC'))
```
ah i see we use here UTC now ... but that doesnt make sense really if we format it as ATOM as ATOM _does_ support timezones.

So either id say we dont use ATOM here or we preserve the timezone and treat this value carefully when reading and convert it to UTC

https://github.com/neos/neos-development-collection/pull/5716#issuecomment-4438715550

nezaniel replied:
Changing the timezone does not change the date, so this would just be just a new convention.
I'm not too invested in this one, though, we can also revert it.

nezaniel replied:
Further idea: how about using the timezone from the CR's clock? That would be a meaningful source of truth

mhsdesign replied:
good one... yes i just want to have the simples option ... and that might very much be UTC all the way. And then as said we dont need to use ATOM as we just put `+00:00` there which when read begs the question if it is always `+00:00` and YES so we could discard that.

mhsdesign replied:
Technically though in a multi user world wide setup an editor would probably like to see _when_ a change was made. And strangely here the timezone of the users e.g. the client could be relevant. But we can just forget that as this is not our fight to fight.

Instead if we wehere to show dates in the ui we could just use the current locale and format the dates accordingly? But then ofc. when the daylightsaving is changed suddenly all previous changes happened one hour off. Is there a solution for this? I really like to just store UTC and dont make it complex 😅

mhsdesign replied:
So in conclusion: The `recordedAt` metadata is stored in the evenstore in a separate datetime column and thus MUST be UTC for our own sanity.

Additionally to the Neos Cr adds custom metadata in a json blob. Here we handle date formatting of "initiatingTimestamp" ourselves and via ATOM we _do_ support timezones.

The question is now which way to go:

- Keep it as in 9.0, write with timezone - read with timezone
- Convert the "initiatingTimestamp" to store UTC - adds obsolete `+00:00` and for legacy events convert timezone during reading via setTimezone() to UTC (what we see here)
- Store time stamp like recordedAt without timezone information to denote that we dont ever want timezones here. Requires workaround for legacy events that stored ATOM. -> e.g count string lenght.

cc @kitsunet

mhsdesign replied:
Okay @kitsunet and me agreed on proceeding with option 2. as this is the best solution in regards that we find out when people might have updated their system to the new UTC change and which events are new. Also this is not as invasive in a bugfix.

mhsdesign committed:
TASK: Simplify Timestamps tests by using UTC

... and test timezones separately

mhsdesign committed:
TASK: Move Timestamps to own namespace - as it's not NodeTraversal

mhsdesign committed:
TASK: Reintroduce testing of timezones

... but use +2 as offset as its better to distinct from of-by-one-errors :O :D

Also extract simple UTC test for non timezone based assertions

mhsdesign committed:
PATCH: Rename to 02

mhsdesign committed:
BUGFIX: Do not rely on 'initiatingTimestamp' being UTC

mhsdesign committed:
PATCH: Respect invalid dates in getInitiatingTimestamp()

mhsdesign commented on Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Projection/NodeRecord.php:
```diff
@@ -147,10 +147,10 @@ public static function createNewInDatabase(
-                'originalcreated' => $timestamps->originalCreated,
-                'lastmodified' => $timestamps->lastModified,
-                'originallastmodified' => $timestamps->originalLastModified,
+                'created' => $timestamps->created->setTimezone(new \DateTimeZone('UTC')),
```
im not super happy with all these explicit `DateTimeZone` mutations before writing.

I think the `TimeStamp` value object could enforce that we always ever deal with UTC and then we dont have to convert here.

Also the conversion creates always a new object instance even there is nothing to do ... this is reluctant 

```
$dateUtc = \DateTimeImmutable::createFromFormat(
    \DateTimeImmutable::ATOM,
    '2024-09-23T13:00:00+00:00'
);

assert($dateUtc === $dateUtc);
assert($dateUtc->getTimezone()->getName() === '+00:00');
// Fails, new instance created
assert($dateUtc === $dateUtc->setTimezone(new \DateTimeZone('+00:00')));
```

mhsdesign replied:
Simplified via https://github.com/neos/neos-development-collection/pull/5716/commits/01eb83d47fcf99120c1c03177e783e35f3c6c48b

mhsdesign committed:
TASK: Simplify `Timestamps` to require to be UTC

The `->setTimezone(new \DateTimeZone('UTC'))` were obsolete now, because they are now UTC. PHP still creates a new instance which is redundant.
 Timestamps should be guaranteed to be UTC

see https://github.com/neos/neos-development-collection/pull/5716#discussion_r3292562406

mhsdesign committed:
TASK: Ensure we use UTC date time when mocking nodes

mhsdesign committed:
TASK: Fix asserted `initiatingTimestamp` in EventExport

mhsdesign committed:
TASK: Add tests to manifest behavior of `InitiatingEventMetadata`

including legacy upcasting

mhsdesign committed:
TASK: Ensure not to create new date time instance for UTC dates
